[OpenCL] Implementation improvements

lukeiwanski opened the pull request:
OpenCL implementation improvements (#22)

- [Build] Use gcc/g++ as a host compiler to avoid #8394 (#54)
- [Eigen] Version bump
- [OpenCL] Implementation improvements
  - Register SYCL implementations for random ops
    - Simplify by using Eigen math functions
  - Registers Scatter and ScatterNd Ops for SYCL
  - Registers Stack op for SYCL
  - Fixes No sycl buffer found error for debug ops
  - Registers MatMul and Transpose Ops to SYCL device for double
  - Extends analyzer_cli_test.py test to cover SYCL
  - Fixes Transpose Op for double when on SYCL
  - Bumps Eigen version to fix double precision issue on SYCL
  - Extends SessionDebugTestBase to cover SYCL
  - Bumps Eigen Version
  - Refactors Ops registration
  - Introduces workaround for Const Op related to the difference between
   CUDA which uses pointers and OpenCL that uses buffers/accessors
  - Extends memory types to cover DEVICE_SYCL as well
  - Introduces  GetSYCLDevice() method that returns list of supported devices
   with GPU device having the highest priority ( doesn't include blacklisted devices )
  - ::internal::Transpose -> tensorflow::internal::Transpose in order to
   avoid compilation reported error
  - Adds sycl_runtime to bazels ARRAY_DEPS
  - Replicates TF_CALL_GPU_PROXY_TYPES for SYCL
  - Fixes an issue caused by switch to aligned allocator for sycl buffer (#53)
  - Fix testSimple and testConst in stack_op_test (#3)
  - RandomGamma has no GPU friendly implementation (#57)
  - Register batch normalization kernels for OpenCL (#61)
  - Compatibility fixes for TensorFlow 1.1.0-rc1
  - Fixes Scatter Op
  - Implements BatchMatmul Op for SYCL
  - Lowercase the device name when GPU or SYCL returned
  - kernel_estimator_test.py assertEqual-> assertAlmostEqual due to floating point 
    representation on the device

Commit:
OpenCL Improvements

* Registers Scatter and ScatterNd Ops for SYCL

* Registers Stack op for SYCL

* Fixes No sycl buffer found error for debug ops

* Registers MatMul and Transpose Ops to SYCL device for double

* Extends analyzer_cli_test.py test to cover SYCL

* Fixes Transpose Op for double when on SYCL

* Bumps Eigen version to fix double precision issue on SYCL

* Extends SessionDebugTestBase to cover SYCL

ville-k committed:
Register SYCL implementations for random ops

Commit:
OpenCL improvements

 - Bumps Eigen Version
 - Refactors Ops registration
 - Introduces workaround for Const Op related to the difference between
   CUDA which uses pointers and OpenCL that uses buffers/accessors
 - Extends memory types to cover DEVICE_SYCL as well
 - Introduces  GetSYCLDevice() method that returns list of supported devices
   with GPU device having the highest priority ( doesn't include blacklisted devices )
 - ::internal::Transpose -> tensorflow::internal::Transpose in order to
   avoid compilation reported error
 - re-introduces fix for bugged string replacement causing a lot of compilation
   warnings -c -> --include
 - Adds sycl_runtime to bazels ARRAY_DEPS
 - Replicates TF_CALL_GPU_PROXY_TYPES for SYCL

ville-k committed:
Avoid functions that might not be defined on SYCL device (#51)

* Avoid functions that might not be defined on SYCL device

* Simplify by using Eigen math functions

Commit:
[OpenCL] Fixes an issue caused by switch to aligned allocator for sycl buffer (#53)

Commit:
[Build] Use gcc/g++ as a host compiler to avoid https://github.com/tensorflow/tensorflow/issues/8394 (#54)

Commit:
[OpenCL] Fixes Scatter Op

Zakor94 committed:
Fix testSimple and testConst in stack_op_test (#3)

* Fix testSimple and testConst in stack_op_test

* Create a specialisation of DoParallelConcatUpdate for SyclDevice and
register it

* Guard all code in TENSORFLOW_USE_SYCL

* Do not use sycl device for int32

* Registration of the Sycl version is now looking like the one for the GPU

* Remove added empty line

Commit:
[OpenCL] Implements BatchMatmul Op for SYCL

Commit:
[OpenCL] RandomGamma has no GPU friendly implementation (#57)

ville-k committed:
Register batch normalization kernels for OpenCL (#61)

Commit:
[OpenCL] Compatibility fixes for TensorFlow 1.1.0-rc1

Commit:
Lowercase the device name when GPU or SYCL returned

Commit:
[OpenCL] kernel_estimator_test.py assertEqual-> assertAlmostEqual due to floating point representation on the device

Commit:
[Eigen] Version bump

tensorflow-jenkins commented:
Can one of the admins verify this patch?

lukeiwanski commented:
@ville-k, @Zakor94 can you confirm?

drpngx commented:
Sometimes googlebot gets confused, could you `git pull --rebase` and push again?

ville-k commented:
@lukeiwanski @googlebot  ok with me.

Zakor94 commented:
@lukeiwanski @googlebot yes ok with me too.

Commit:
Merge remote-tracking branch 'upstream/master' into opencl_improvements

Commit:
GPU device name string manipulation is not needed anymore

Commit:
[OpenCL] Adds SYCL to device backwards compatibility

Commit:
[OpenCL] Extends core_rnn_test.py to run for SYCL device

Zakor94 committed:
[OpenCL] Minor optimizations for build script

Zakor94 committed:
[OpenCL] Enables skip folder list in build script

Commit:
Merge branch 'master' into opencl_improvements

Commit:
[OpenCL] Fixes ApplyAdamOp for Sycl device

Commit:
[OpenCL] SYCL device improvements

Commit:
[OpenCL] Fixes debug_ops's SEGFAULT for SYCL device

Commit:
[Build] Adds hexagon to skipped folders list

Commit:
[OpenCL] Removes EnterLameDuckMode from SYCL device and allocator

Commit:
[OpenCL] Registers Unique Op for SYCL device

Commit:
[OpenCL][Temporary] Disables tests for SYCL target due to features not being implemented yet

  Tests affected:
    - tensorflow/contrib/memory_stats/python/kernel_tests/memory_stats_ops_test.py
    - tensorflow/contrib/rnn/python/kernel_tests/core_rnn_test.py
    - tensorflow/python/kernel_tests/conv_ops_test.py
    - tensorflow/python/kernel_tests/depthwise_conv_op_test.py
    - tensorflow/python/kernel_tests/pooling_ops_3d_test.py
    - tensorflow/python/kernel_tests/pooling_ops_test.py
    - tensorflow/python/kernel_tests/scatter_nd_ops_test.py
    - tensorflow/python/training/adam_test.py
    - tensorflow/python/training/localhost_cluster_performance_test.py
    - tensorflow/python/training/training_ops_test.py

Commit:
[OpenCL][Temporary] Disables failing tests for SYCL in order to establish regression baseline

  Tests affected:
    - tensorflow/python/debug/cli/analyzer_cli_test.py
    - tensorflow/python/debug/lib/session_debug_testlib.py
    - tensorflow/python/debug/lib/stepper_test.py
    - tensorflow/python/kernel_tests/unstack_op_test.py
    - tensorflow/python/ops/image_ops_test.py

Commit:
[OpenCL] Take options.config.device_count() into consideration

Commit:
[OpenCL] Fixes compilation warning

Commit:
[OpenCL] device:SYCL:0 -> sycl:0

Zakor94 committed:
[OpenCL] Removes unwanted flags in building script

Removes flags given to computecpp that enable SIMD instructions
Removes duplicate flags

Commit:
bool -> const bool

Commit:
[OpenCL] sycl in test_util.gpu_device_name() -> is_sycl_enabled()

Commit:
Merge branch 'master' into opencl_improvements

Commit:
[OpenCL][Temporary] Disables failing tests for SYCL in order to establish regression baseline

  Test affected:
    - tensorflow/contrib/stateless/python/kernel_tests/stateless_random_ops_test.py

Commit:
Imports test_util from tensorflow.python.framework

yifeif commented:
@tensorflow-jenkins test this please

Commit:
[OpenCL] Fixes formatting in Python code

Commit:
[OpenCL] Extends session_test.py to cover SYCL device

yifeif commented:
@tensorflow-jenkins test this please

Commit:
[OpenCL] Cleans singleton class

yifeif commented:
@tensorflow-jenkins test this please

Commit:
[OpenCL] Keeping CUDA happy

Commit:
Merge remote-tracking branch 'upstream/master' into opencl_improvements

Commit:
[OpenCL][Temporary] Disables failing tests for SYCL in order to establish regression baseline

  Test affected:
   - tensorflow/contrib/rnn/python/kernel_tests/core_rnn_cell_test.py
   - tensorflow/contrib/seq2seq/python/kernel_tests/beam_search_ops_test.py

yifeif commented:
@tensorflow-jenkins test this please

alistair-low committed:
Added support for building with SYCL on ARM.

drpngx commented:
Jenkins, test this please.

Rbiessy committed:
[OpenCL] Fixes scatter_nd_op_test

benoitsteiner commented on tensorflow/contrib/memory_stats/python/kernel_tests/memory_stats_ops_test.py:
```diff
@@ -32,7 +32,7 @@ class MemoryStatsOpsTest(test_util.TensorFlowTestCase):
     # AllocatorStats.bytes_limit is set to zero for CPU allocators, so we skip
     # the check.
-    if not test.is_gpu_available():
+    if not test.is_gpu_available() or test_util.is_sycl_enabled():
```
Please delete all the test_util.is_sycl_enabled(). The point of having test is to expose problems, so hiding them is defeating the purpose.

lukeiwanski replied:
Removed in https://github.com/tensorflow/tensorflow/pull/9117/commits/73e405572a59ba8ab1d92dc3f7cddff5c9fb240c

The intention behind disabling the failing tests was to establish a regression baseline so that we can establish a smooth development process. We would enable more tests as we add the corresponding features, while being able to continually test that we haven't introduced regressions.

benoitsteiner commented on tensorflow/core/common_runtime/direct_session_with_tracking_alloc_test.cc:
```diff
@@ -138,7 +138,8 @@ TEST(DirectSessionWithTrackingAllocTest, CostModelWarmup) {
   CostModelManager::CostModelMap cost_models;
   ds->ExportCostModels(&cost_models);
-  CHECK_EQ(cost_models.size(), 1);
+  ASSERT_GE(2, cost_models.size());
```
Why is this needed ?

lukeiwanski replied:
Fixed in: https://github.com/tensorflow/tensorflow/pull/9117/commits/73e405572a59ba8ab1d92dc3f7cddff5c9fb240c 

That should not be there. Auto merge conflict resolution cannot be trusted.

Commit:
Acts on the review feedback from:
 - https://github.com/tensorflow/tensorflow/pull/9117#discussion_r113608975
 - https://github.com/tensorflow/tensorflow/pull/9117#discussion_r113609173

Commit:
Fixes auto-merge mistake

drpngx commented:
Jenins, test this please.

vrv commented:
@lukeiwanski these massive PRs are hard to review since there is a stream of new commits being added, and the PR is already very large.

We typically try to stage these changes in small PRs that are more easily reviewable -- can you break this change into smaller requests in some way?  As it is, I don't see anybody being able to safely review a change of this magnitude (117 files!)

lukeiwanski commented:
@vrv Ok I will try. What is desired changeset size?

Commit:
[OpenCL] struct SyclDevice -> class SyclDevice

Commit:
Revert "[OpenCL] struct SyclDevice -> class SyclDevice"

This reverts commit addd43348c374a5379f67bb1e5ad084715722fc2.

vrv commented:
Benoit would have a better idea of what he'd like, but it's much easier to review a change that has like, <10 files, < 200-300 LoC change at a time; that we he can review it, it will have fewer things to change, and we can make more iterative progress.

Might be worth you and Benoit talking offline to see what's going to work well!

(I'm just on sync rotation this week and trying to help move things along :)

lukeiwanski commented:
Appriciate and thanks @vrv !
My plan for this PR is to revert refactoring changes ( that should knock off 77 files ) and re-introduce them in a new one ( or series ) - I will check with @benoitsteiner on how to proceed.

Commit:
[OpenCL] Reverting refactoring commit.

  As requested in the review https://github.com/tensorflow/tensorflow/pull/9117#issuecomment-298454466
  This change set will be re-introduced in smaller chunks.

lukeiwanski commented:
@vrv Is current size of the pull requst acceptable or should I keep spliting it?

benoitsteiner commented on tensorflow/core/common_runtime/sycl/sycl_device_factory.cc:
```diff
@@ -18,24 +18,34 @@ limitations under the License.
+
     for (int i = 0; i < n; i++) {
-      string name = strings::StrCat(name_prefix, "/device:SYCL:", i);
+      string name = strings::StrCat(name_prefix, "/sycl:", i);
```
Can yo revert this change ? All devices should be named /device:DEVICE_NAME:DEVICE_ID (with the exception of CPUs and GPUs, which can also be referred to as /cpu:cpu_id and /gpu:gpu_id in order to preserve backward compatibility).

lukeiwanski replied:
can DEVICE_NAME be lower case?

lukeiwanski replied:
Fixed in https://github.com/lukeiwanski/tensorflow/commit/c4b1480529f03f17f16914d15e0e0576d480cf57

benoitsteiner commented on tensorflow/core/kernels/constant_op.cc:
```diff
@@ -184,6 +210,43 @@ class FillOp : public OpKernel {
+    T data_host;
+    auto device = context->eigen_sycl_device();
+    auto size = sizeof(T);
+    auto src_ptr = GetBase(&Tvalue);
```
Handling of device pointers should really be done by the sycl compiler and in eigen, in order to avoid opening a can or worms in TensorFlow.

lukeiwanski replied:
We are aware of the issue and working on resolving it. I removed above workaround in https://github.com/lukeiwanski/tensorflow/commit/1a1525bae9a7e57a3faa2ed8e8a70de9dff3747f where I am registering Fill operation with .HostMemory("value"). Is that acceptable for time being?

lukeiwanski replied:
8ba7dbdbcf9dc1e540a47988437093e2d2fd2420 seems to be more appropriate solution

Commit:
Revert "[OpenCL] device:SYCL:0 -> sycl:0"

This reverts commit cf16e60340b62d16c3764d71b716fe03d35f87a9.

Commit:
Revert "[OpenCL] Adds SYCL to device backwards compatibility"

This reverts commit b8401b5164199b7a169be1c1d8dea5001195c390.

Commit:
Acts on the feedback from https://github.com/tensorflow/tensorflow/pull/9117#discussion_r115036905

Commit:
control_flow_ops_py_test.py expects device name to be lower cased

Commit:
Acts on the feedback from https://github.com/tensorflow/tensorflow/pull/9117#discussion_r115037222

Commit:
Removes debug print

Commit:
Removes not needed partial specialisation

Commit:
[OpenCL] Registers ScatterNdFunctor for SYCL device

tensorflow-jenkins commented:
Can one of the admins verify this patch?

Commit:
Merge remote-tracking branch 'upstream/master' into opencl_improvements

Commit:
Merge remote-tracking branch 'upstream/master' into opencl_improvements

Rbiessy committed:
[OpenCL] Make it compile

Rbiessy committed:
[OpenCL] Follow gpu_device changes

rmlarsen commented:
@benoitsteiner could you please take another look?

Commit:
[OpenCL] Adds cxx_builtin_include_directory for python lib

  Fixes bazels missing undeclared inclusions that appeared after
  merge with TensorFlow upstream

rmlarsen commented:
@tensorflow-jenkins test this please

lukeiwanski commented:
The MacOs CPU Tests seems to be failing due to a missing dependency in gif_archive..

ERROR: /private/var/tmp/_bazel_jenkins/13e370a18c169b19baeafefb05212b85/external/gif_archive/BUILD.bazel:8:1: undeclared inclusion(s) in rule '@gif_archive//:gif':
this rule is missing dependency declarations for the following files included by 'external/gif_archive/lib/openbsd-reallocarray.c':
 
Is that caused by our changeset or is it present in the upstream? I cannot see how we could cause it.

rmlarsen commented:
@lukeiwanski no, the build error on MacOS is unrelated to your changes.

Commit:
[OpenCL] Fixes Constant Op

gunan commented on tensorflow/contrib/kernel_methods/python/kernel_estimators_test.py:
```diff
@@ -92,7 +92,7 @@ def testLinearlySeparableBinaryDataNoKernels(self):
     # loss and perfect accuracy.
     self.assertLess(metrics['loss'], 0.1)
-    self.assertEqual(metrics['accuracy'], 1.0)
+    self.assertAlmostEqual(metrics['accuracy'], 1.0, 7)
```
According to the comment above, I am worried about almost equal here. I know 7 digits is very very precise, but still we are making the test looser.

lukeiwanski replied:
The reason for that was the fact that sometimes OpenCL represent this value as 0.99999996
Reverting in 1a6045241a27228ba817cb0b722679c357077e36

gunan commented:
Jenkins, test this please.

Commit:
[OpenCL] gXX-4.8 -> gXX

Commit:
[OpenCL] Removes -D_GLIBCXX_USE_CXX11_ABI=0 as it breaks default compiler setup for Ubuntu 16.04

lukeiwanski commented on configure:
```diff
@@ -683,7 +683,7 @@ if [ "$TF_NEED_OPENCL" == "1" ]; then
   fromuser=""
   if [ -z "$HOST_CXX_COMPILER" ]; then
-    default_cxx_host_compiler=$(which clang++-3.6 || true)
+    default_cxx_host_compiler=$(which g++-4.8 || true)
```
We locked to GCC-4.8 as it was default installation on Ubuntu 14.04.
This has been changed in 50197b9792e3eb262201ee143cc82546a151e388 and d777320c6a9b730449f744a5ad8d068da12953ed
However, on ubuntu 16.04 user will have to add "-D_GLIBCXX_USE_CXX11_ABI=0" manually to the command line if there is custom compiler installed on the machine.

Commit:
Revert "[OpenCL] kernel_estimator_test.py assertEqual-> assertAlmostEqual due to floating point representation on the device"

This reverts commit 06c50c0a485f40c30a436f02c3fa7794e370c49d.

Commit:
[OpenCL] CPU allocator is a singleton we should not delete it

benoitsteiner commented:
Jenkins, test this please.
